File: filefn/server/src/dedup/service.ts
import type { Adapter } from '@superfunctions/db';
import { createHash } from 'node:crypto';
import { type StorageAdapter, type StorageTargetName } from '@superfunctions/storage';
import { resolveStorageTarget, type PolicyRegistry } from '../policies.js';

export interface DeduplicationServiceConfig {
  db: Adapter;
  policies?: Pick<PolicyRegistry, 'get'>;
  namespace?: string;
  enabled?: boolean;
}

export interface FileVersionRecord {
  versionId: string;
  fileId: string;
  storageKey: string;
  mimeType: string;
  size: number;
  checksumSha256Base64?: string | null;
  tenantId?: string | null;
  createdAt: string;
}

export interface DeduplicationResult {
  isDuplicate: boolean;
  existingVersionId?: string;
  existingStorageKey?: string;
  checksumSha256Base64: string;
}

interface FileRecord {
  fileId: string;
  policy: string;
}

export function createDeduplicationService(config: DeduplicationServiceConfig) {
  const { db, policies, namespace = 'filefn', enabled = true } = config;

  async function resolveVersionStorageTarget(version: Pick<FileVersionRecord, 'fileId'>): Promise<string> {
    const file = await db.findOne<FileRecord>({
      model: 'files',
      where: [{ field: 'fileId', operator: 'eq', value: version.fileId }],
      namespace,
    });
    const policy = file ? policies?.get(file.policy) : undefined;
    return resolveStorageTarget(policy ?? {});
  }

  function computeHash(data: Uint8Array): string {
    const hash = createHash('sha256');
    hash.update(data);
    return hash.digest('base64');
  }

  async function computeHashFromStream(stream: ReadableStream<Uint8Array>): Promise<string> {
    const hash = createHash('sha256');
    const reader = stream.getReader();

    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        if (value) {
          hash.update(value);
        }
      }
    } finally {
      reader.releaseLock();
    }

    return hash.digest('base64');
  }

  async function checkForDuplicate(
    checksumSha256Base64: string,
    tenantId: string | null,
    storageTarget?: StorageTargetName | null,
  ): Promise<DeduplicationResult> {
    if (!enabled) {
      return {
        isDuplicate: false,
        checksumSha256Base64,
      };
    }

    const whereConditions: Array<{
      field: string;
      operator: 'eq' | 'ne' | 'gt' | 'gte' | 'lt' | 'lte' | 'in' | 'not_in' | 'contains' | 'starts_with' | 'ends_with';
      value: any;
    }> = [{ field: 'checksumSha256Base64', operator: 'eq', value: checksumSha256Base64 }];

    if (tenantId !== null) {
      whereConditions.push({ field: 'tenantId', operator: 'eq', value: tenantId });
    } else {
      whereConditions.push({ field: 'tenantId', operator: 'eq', value: null });
    }

    if (!storageTarget) {
      const existingVersion = await db.findOne<FileVersionRecord>({
        model: 'fileVersions',
        where: whereConditions,
        namespace,
      });

      if (existingVersion) {
        return {
          isDuplicate: true,
          existingVersionId: existingVersion.versionId,
          existingStorageKey: existingVersion.storageKey,
          checksumSha256Base64,
        };
      }

      return {
        isDuplicate: false,
        checksumSha256Base64,
      };
    }

    const existingVersions = await db.findMany<FileVersionRecord>({
      model: 'fileVersions',
      where: whereConditions,
      namespace,
    });

    for (const existingVersion of existingVersions) {
      if (storageTarget) {
        const existingTarget = await resolveVersionStorageTarget(existingVersion);
        if (existingTarget !== storageTarget) {
          continue;
        }
      }

      return {
        isDuplicate: true,
        existingVersionId: existingVersion.versionId,
        existingStorageKey: existingVersion.storageKey,
        checksumSha256Base64,
      };
    }

    return {
      isDuplicate: false,
      checksumSha256Base64,
    };
  }

  return {
    isEnabled(): boolean {
      return enabled;
    },

    /**
     * Computes SHA-256 hash of the provided data.
     * Returns base64-encoded hash.
     */
    computeHash,

    /**
     * Computes SHA-256 hash from a readable stream.
     * Returns base64-encoded hash.
     */
    computeHashFromStream,

    /**
     * Checks if a version with the given hash already exists in the same tenant.
     * Returns deduplication result with existing version info if found.
     */
    checkForDuplicate,

    /**
     * Computes hash from stored content and checks for duplicates.
     * This is used during upload completion.
     */
    async computeAndCheckDuplicate(
      storageKey: string,
      tenantId: string | null,
      storageTargetOrStorage: StorageTargetName | null | Pick<StorageAdapter, 'openDownloadStream'>,
      maybeStorage?: Pick<StorageAdapter, 'openDownloadStream'>
    ): Promise<DeduplicationResult> {
      const storageTarget =
        maybeStorage === undefined
          ? null
          : storageTargetOrStorage as StorageTargetName | null;
      const storage =
        maybeStorage ??
        (storageTargetOrStorage as Pick<StorageAdapter, 'openDownloadStream'>);

      if (!storage.openDownloadStream) {
        throw new Error('Storage adapter does not support streaming downloads required for deduplication');
      }

      const stream = await storage.openDownloadStream({ key: storageKey, target: storageTarget ?? undefined });
      const checksumSha256Base64 = await computeHashFromStream(stream);

      if (!enabled) {
        return {
          isDuplicate: false,
          checksumSha256Base64,
        };
      }

      return checkForDuplicate(checksumSha256Base64, tenantId, storageTarget);
    },

    /**
     * Verifies that a hash matches the content at the given storage key.
     * Used for integrity verification.
     */
    async verifyHash(
      storageKey: string,
      expectedHash: string,
      storage: Pick<StorageAdapter, 'openDownloadStream'>,
      storageTarget?: StorageTargetName | null,
    ): Promise<boolean> {
      if (!storage.openDownloadStream) {
        return false;
      }

      const stream = await storage.openDownloadStream({ key: storageKey, target: storageTarget ?? undefined });
      const actualHash = await computeHashFromStream(stream);

      return actualHash === expectedHash;
    },
  };
}

export type DeduplicationService = ReturnType<typeof createDeduplicationService>;

File: filefn/python/filefn/server/dedup/service.py
import hashlib
import base64
from typing import Any, Optional, AsyncGenerator, Dict, Protocol
from pydantic import BaseModel
from superfunctions.db import Adapter

from ..policies import resolve_storage_target

class DeduplicationServiceConfig(BaseModel):
    db: Any # Adapter
    policies: Optional[Any] = None
    namespace: str = 'filefn'
    enabled: bool = True
    
    class Config:
        arbitrary_types_allowed = True

class DeduplicationResult(BaseModel):
    isDuplicate: bool
    existingVersionId: Optional[str] = None
    existingStorageKey: Optional[str] = None
    checksumSha256Base64: str

class DeduplicationService:
    def __init__(self, config: DeduplicationServiceConfig):
        self.db = config.db
        self.policies = config.policies
        self.namespace = config.namespace
        self.enabled = config.enabled

    async def _resolve_version_storage_target(self, file_id: str) -> str:
        file_row = await self.db.find_one(
            model='files',
            where=[{'field': 'fileId', 'operator': 'eq', 'value': file_id}],
            namespace=self.namespace,
        )
        policy_name = file_row.get('policy') if file_row else None
        policy = self.policies.get(policy_name) if (self.policies and policy_name) else None
        return resolve_storage_target(policy)

    def is_enabled(self) -> bool:
        return self.enabled

    def compute_hash(self, data: bytes) -> str:
        sha256 = hashlib.sha256()
        sha256.update(data)
        return base64.b64encode(sha256.digest()).decode('utf-8')

    async def compute_hash_from_stream(self, stream: AsyncGenerator[bytes, None]) -> str:
        sha256 = hashlib.sha256()
        async for chunk in stream:
            sha256.update(chunk)
        return base64.b64encode(sha256.digest()).decode('utf-8')

    async def check_for_duplicate(
        self,
        checksum_sha256_base64: str,
        tenant_id: Optional[str],
        storage_target: Optional[str] = None,
    ) -> DeduplicationResult:
        if not self.enabled:
            return DeduplicationResult(
                isDuplicate=False,
                checksumSha256Base64=checksum_sha256_base64
            )

        where_conditions = [
            {'field': 'checksumSha256Base64', 'operator': 'eq', 'value': checksum_sha256_base64}
        ]

        if tenant_id:
            where_conditions.append({'field': 'tenantId', 'operator': 'eq', 'value': tenant_id})
        else:
             where_conditions.append({'field': 'tenantId', 'operator': 'eq', 'value': None})

        existing_versions = await self.db.find_many(
            model='fileVersions',
            where=where_conditions,
            namespace=self.namespace
        )

        for existing_version in existing_versions:
            if storage_target is not None:
                existing_target = await self._resolve_version_storage_target(existing_version.get('fileId'))
                if existing_target != storage_target:
                    continue
            return DeduplicationResult(
                isDuplicate=True,
                existingVersionId=existing_version.get('versionId'),
                existingStorageKey=existing_version.get('storageKey'),
                checksumSha256Base64=checksum_sha256_base64
            )

        return DeduplicationResult(
            isDuplicate=False,
            checksumSha256Base64=checksum_sha256_base64
        )

    async def compute_and_check_duplicate(
        self,
        storage_key: str,
        tenant_id: Optional[str],
        storage_target: Optional[str],
        storage: Any,
    ) -> DeduplicationResult:
        if not self.enabled:
             return DeduplicationResult(
                isDuplicate=False,
                checksumSha256Base64=''
            )

        if not hasattr(storage, 'open_download_stream'):
            raise Exception('Storage adapter does not support streaming downloads required for deduplication')

        stream = await storage.open_download_stream(key=storage_key, target=storage_target)
        checksum = await self.compute_hash_from_stream(stream)
        
        return await self.check_for_duplicate(checksum, tenant_id, storage_target)

    async def verify_hash(self, storage_key: str, expected_hash: str, storage: Any, storage_target: Optional[str] = None) -> bool:
        if not hasattr(storage, 'open_download_stream'):
            return False
            
        stream = await storage.open_download_stream(key=storage_key, target=storage_target)
        actual_hash = await self.compute_hash_from_stream(stream)
        
        return actual_hash == expected_hash

def create_deduplication_service(config: DeduplicationServiceConfig) -> DeduplicationService:
    return DeduplicationService(config)
